src/python/pants/engine/scheduler.py

```diff
@@ -487,12 +487,13 @@ def products_request(self, products, subjects):
         cumulative_trace = '\n'.join(self.trace(request))
         raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
 
-      if len(throw_root_states) == 1:
+      unique_exceptions = set(t.exc for t in throw_root_states)
+      if len(unique_exceptions) == 1:
         raise throw_root_states[0].exc
       else:
         raise ExecutionError('Multiple exceptions encountered:\n  {}'
-                             .format('\n  '.join('{}: {}'.format(type(t.exc).__name__, str(t.exc))
-                                                 for t in throw_root_states)))
+                             .format('\n  '.join('{}: {}'.format(type(t).__name__, str(t))
```

Comment:
nit: This makes the order of exception reports non-deterministic. That's probably fine (and I expect the engine probably reports them in a non-deterministic order, too), but it may be worth preserving their reported order. Probably not, though :)

Reply:
That's fair. I feel like what we need is to have a better approach to reporting errors in general, but I'm still not sure what that looks like. That's why there's a TODO above and #3912. The specific issues in #3912 are better, but not yet good. 

The errors appear to be deterministic when you look at them with the full trace, so at least theres that.

Reply:
could you just throw in a `sorted(unique_exceptions)` here, maybe?